src/extension/cli/server.ts

```diff
@@ -0,0 +1,86 @@
+
+import { ChildProcessWithoutNullStreams, spawn } from 'node:child_process'
+import path from 'node:path'
+import fs from 'node:fs/promises'
+
+import vscode, { Disposable } from 'vscode'
+
+import { SERVER_ADDRESS } from '../../constants'
+
+import ServerError from './serverError'
+
+export interface IServerConfig {
+    assignPortDynamically?: boolean
+    retryOnFailure?: boolean
+    maxNumberOfIntents: number
+}
+
+class Server implements Disposable {
+
+    #runningPort: number
+    #process: ChildProcessWithoutNullStreams | undefined
+    #binaryPath: string
+    #retryOnFailure: boolean
+    #intent: number
+    #maxNumberOfIntents: number
+
+    constructor(options: IServerConfig) {
+        const config = vscode.workspace.getConfiguration('runme.server')
+        this.#runningPort = config.get<number>('port')!
+        this.#binaryPath = path.join(__dirname, '../', config.get<string>('binaryPath')!)
+        this.#retryOnFailure = options.retryOnFailure || false
+        this.#intent = 0
+        this.#maxNumberOfIntents = options.maxNumberOfIntents
+    }
+
+    dispose() {
+        this.#process?.removeAllListeners()
+        this.#process?.kill()
+    }
+
+    async #start(): Promise<object | ServerError> {
+        this.#intent ++
+        if (this.#maxNumberOfIntents < this.#intent) {
+            return new ServerError(`Cannot start server, reached max number of intents: ${this.#maxNumberOfIntents}`)
+        }
+        const binaryExists = await fs.access(this.#binaryPath)
+            .then(() => true, () => false)
+        if (!binaryExists) {
+            throw new ServerError('Cannot find server binary file')
+        }
+        return new Promise((resolve, reject) => {
+            this.#process = spawn(this.#binaryPath, [
+                'server',
+                '--address',
+                `${SERVER_ADDRESS}:${this.#runningPort}`
+            ])
+            this.#process.stderr.on('data', (data) => {
+                // TODO: Parse data to detect if there is any error, for now, simulating nothing failed.
+                const isError = false
+                if (isError) {
+                    return reject(new ServerError(`Server failed, reason: ${data.toString()}`))
+                }
+
+                const msg = data.toString()
+                try {
+                  const log = JSON.parse(msg)
+                  if ('started listening' === log?.['msg']) {
+                    this.#retryOnFailure && isError ?
+                        this.#start() :
+                        resolve(log)
+                  }
+                } catch (err: any) {
+                  reject(new ServerError(msg))
```

Comment:
As babik noted, non-runme messages can be logged in any format, and those may not necessarily be unrecoverable errors. I don't think you want to reject if there's a parse error, maybe just log it to console.

Reply:
Changed to the specific use case when rejecting makes sense (when starting the server) other cases will be logged (under a Runme setting flag)